README.md
# YouthRx
Scrape and clean Sepohora products and ingredients data

The scraper is based on Jennifer Ruddock's [article](https://nycdatascience.com/blog/student-works/scraping-sephora-an-ingredients-analysis/) and [repo](https://github.com/chemjen/skincare_scraping)

To run scraper: Run get_brand_list.py then get_products_list.py then sephora.py


changelog.txt
get_brands_lists.py:
- line 18:
    change brands = driver.find_elements_by_xpath('//div//ul/li/a')
        to brands = driver.find_elements_by_xpath('//a[@data-at="brand_link"]')
    problem: the elements of the website has changed since the author created this scraper. It grabbed the products link instead of brand link.
    solution: grab any elements with brand_link in their data-at attribute

get_products_list.py:
- line 58:
    replace Mozilla/5.0 (Macintosh; Intel Mac OS X \ 10_13_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 \ Safari/537.36
       with Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36
  find your user agent at https://www.whatsmyua.info/
- line 69:
    change @id to @data-at
    reason: website changed their element attribute
- line 142: add driver.implicitly_wait(10). Force webdriver to wait 10 secs to find the element
    reason: driver couldn't find element because it wasn't visible
- line 122: add
    wait = WebDriverWait(driver, 10)
    element = wait.until(EC.element_to_be_clickable(ex_out))
- line 157: add
    wait = WebDriverWait(driver, 10)
    element = wait.until(EC.element_to_be_clickable(ex_out))

Scraper/sephora.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.keys import Keys
from fake_useragent import UserAgent

import time
import random
import re
import os
import csv
import math

### This script is for scraping the sephora product pages - getting
### information on each product:

### name & brand: the name and brand of the product

### The family/genus/species of the product,
### representing the hierarchy of categorization: "family" is the  main
### category (Makeup, Hair, Skincare, Men, or  Fragrance), the "genus" is the
### category within the family, and the "species" is the subcategory within
### the genus.

### price: The full price (USD) of the product, for its default size and/or
### color
### weight & volume: the weight (oz) and volume (mL) of the product, for its
### default size

### num_loves: the number of "loves" the product has received

### num_reviews: the number of reviews the product has
### if the number of reviews is greater than 1000, the nearest floor 100 is
### given at the top of the page. If the number of reviews is greater than
### 10,000, it is given as the nearest floor 1000. For example, if the actual
### number of reviews is 1200, the top of the page will say "1.2K". If the
### number of reviews is 21897, the top of the page will say "21K".
### Similarly, products with millions of reviews were rounded to the nearest
### 10th of a million

### ave_rating: the average rating of the product
### at the top of the page, the value given for the ave_rating is rounded to
### the nearest 0.5. The highest rating is 5 (5 stars)

### details: all the text under the "details" tab
### ingredients: all the text under the "ingredients" tab

##############################################################################
## Setting up the Selenium webdriver and also grabbing the product urls
## which are the result of get_product_urls.py and the Scrapy spider
## and get_chemical_urls.py

ua = UserAgent()
userAgent = ua.random

opts = Options()
user_agent = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.50 Safari/537.36'
opts.add_argument("--disable-notifications")
opts.add_argument(f'user-agent={userAgent}')
# opts.add_argument("--headless")
#opts.add_argument('--disable-gpu')
#opts.add_argument('--no-sandbox')
#opts.add_experimental_option("excludeSwitches", ["enable-automation"])
#opts.add_experimental_option('useAutomationExtension', False)

driver = webdriver.Chrome(options=opts)

driver.delete_all_cookies()

product_urls = open('sephora_product_urls.txt', 'r').readlines()
# product_urls = open('../sephora_chemical_urls.txt', 'r').readlines()

##############################################################################
## This script is designed so that should the program fail for whatever
## reason -- usually because of accidentally quitting the browser window or
## the browser crashing, or a loss of connection, etc -- the current data will
## save so that the driver can be restarted where it left off.

## If the script has not been run before, a "products.csv" file will be
## created with the data written to it. If the script has been run before but
## quit unexpectedly, the index of where it left off will be saved in
## "index.txt", and upon rerunning the script, the products.csv file will be
## openned and appended to, starting from the index in index.txt.

columns = ['name', 'brand', 'family', 'genus', 'species', 'price','ave_rating', 'volume', 'num_loves', 'num_reviews', 'weight', 'details','ingredients','url']
## for writing the first row of the csv using DictWriter
column_dict = dict(zip(columns, columns))

if os.path.isfile(f'index.txt') and os.path.isfile(f'products.csv'):
    index = int(open('index.txt').read())
    csv_file = open(f'products.csv', 'a', encoding='utf-8', newline='')
    writer = csv.DictWriter(csv_file, fieldnames=columns)
    product_urls = product_urls[index:]
else:
    index = 0
    csv_file = open(f'products.csv', 'w', encoding='utf-8', newline='')
    writer = csv.DictWriter(csv_file, fieldnames=columns)
    writer.writerow(column_dict)

##############################################################################

## I chose not to scrape things that were not chemical based (do not have
## ingredients lists) and I also decided not to do anything travel-size or
## mini size to avoid redundant products and skewed price ranges
unwanted_products = ['Mini Size', 'Value & Gift Sets', 'Facial Rollers',
                     'Brushes & Applicators', 'False Eyelashes', 'Rollerballs & Travel Size',
                     'Candles & Home Scents', 'Beauty Supplements', 'High Tech Tools',
                     'Lip Sets', 'Face Sets', 'Eye Sets', 'Makeup Bags & Travel Accessories',
                     'Tweezers & Eyebrow Tools', 'Blotting Papers', 'Hair Tools']

## The xpath for the "continue shopping" button on the pop-up modal dialog
continue_shopping_xpath = '//button[@aria-label="Continue shopping"]'
try:
    for i, url in enumerate(product_urls):
        # keep track of how many pages have been scraped
        print(i + index)
        product_dict = {}
        driver.implicitly_wait(10)
        driver.get(url)
        if i == 0:
            try:
                button = driver.find_element_by_xpath(continue_shopping_xpath)
                button.click()
            except:
                time.sleep(1)
        # product_type is the text containing family/genus/species
        product_type = driver.find_elements_by_xpath('//nav[@aria-label="Br' + \
                                                     'eadcrumbs"]//a')
        # if len(product_type) <  3, it is gift card or gift set of some sort
        print("1")
        driver.save_screenshot("screenshot.png")

        if len(product_type) < 3: continue

        print("2")
        # separate the family, genus and species
        product_type = [val.text for val in product_type]
        # remove products that are in the unwanted_products categories
        if (product_type[1] in unwanted_products) or \
                (product_type[2] in unwanted_products):
            continue
        print("3")
        # get the product name and brand, which is usually up high on the
        # page, but sometimes is more in the center of the page
        product = driver.find_elements_by_xpath('//span[@data-at="product_name"]')
        brand = driver.find_elements_by_xpath('//a[@data-at="brand_name"]')
        if len(product) == 0:
            product = driver.find_elements_by_xpath('//h1[@data-comp="Displ' + \
                                                    'ayName Flex Box"]//span')

        ## add the various values to the product dictionary
        product_dict['url'] = driver.current_url
        product_dict['family'] = product_type[0]
        product_dict['genus'] = product_type[1]
        product_dict['species'] = product_type[2]
        product_dict['name'] = product[0].text
        product_dict['brand'] = brand[0].text
        ## Likewise, the price can be found in two different locations
        product_dict['price'] = driver.find_element_by_xpath('//p[@data-comp="Price "]/span/span[1]/b').text

        ## The size of the product and item number are often listed together
        ## as "XX.XX( fl) oz/XX mL ITEM # XXXXXXXXXX"
        ## sometimes, that xpath only contains the element number, in which
        ##	case the weight & volume are in a different location. Sometimes no
        ## weight/volume are listed
        try:
            size = driver.find_element_by_xpath('//div[@data-comp="SwatchDescription "]').text
        except:
            size=''

        ## once the size variable is found, separate the weight (in oz) from
        ## the volume (in mL)
        ## the weight can also be in fluid oz, but oz and fl oz are treated
        ## the same here
        weight = re.findall("((\d+\.?\d*)( fl)? ?oz)", size)
        if weight:
            product_dict['weight'] = weight[0][1]
        volume = re.findall("(\d+) ?mL", size)
        if volume:
            product_dict['volume'] = volume[0]
        ## get the number of "loves" from the top of the page
        product_dict['num_loves'] = driver.find_element_by_xpath('//div[@data-comp="LovesCount "]/span').text
        ## the number of reviews is only listed if there are reviews
        try:
            product_dict['num_reviews'] = driver.find_element_by_xpath('//span[@data-at="number_of_reviews"]').text.split()[0]
        except:
            product_dict['num_reviews'] = '0'
        # if there are reviews, then get the average review
        if product_dict['num_reviews'] != '0':
            product_dict['ave_rating'] = driver.find_element_by_xpath('//div[@data-comp="StarRating "]').get_attribute('aria-label').split()[0]
        time.sleep(2)

        ## get ingredients and details
        try:
            driver.find_element_by_xpath('//button[@data-at="ingredients"]').click()
            product_dict['ingredients'] = driver.find_element_by_xpath('//div[@id="ingredients"]/div/div').get_attribute("innerHTML")
        except:
            pass

        try:
            driver.find_element_by_xpath('//button[@class="css-5fs8cb eanm77i0"]').click()
            product_dict['details'] = driver.find_element_by_xpath('//div[@class="css-u2scbn eanm77i0"]/div').get_attribute("innerHTML")
        except:
            pass


        ## The number of reviews written at the top of the page tends to be
        ## rounded, as described above. The average rating at the top of the
        ## page also tends to be rounded to the nearest 0.5. However, the
        ## actual average rating and number of reviews is written at the
        ## bottom of the page above the reviews. The driver will attempt to
        ## get the actual number of reviews and the actual average rating, but
        ## if the page length is atypical, the driver might not see those
        ## elements. If the driver does not see them, then it just moves on
        ## with the rounded values from the top of the page. If it does see
        ## them, then it will replace the rounded values from the top of the
        ## page with the actual values from the bottom of the page.
        ## This part of the script could maybe be written better, but in the
        ## interest of time and efficiency, I chose to move on.
        driver.execute_script("arguments[0].scrollIntoView();", driver.find_element_by_xpath('//div[@id="ratings-reviews-container"]'))

        if product_dict['num_reviews'] != '0':
            try:
                ave_rating = driver.find_element_by_xpath('//span[@class="css-1ac1x0l eanm77i0"]').get_attribute("innerHTML")
                product_dict['ave_rating'] = ave_rating

            except:
                pass
            try:
                num_reviews = driver.find_element_by_xpath('//div[contains(@data-comp,"ReviewsStats")]//span[@class="css-nv7myq eanm77i0"]').text.split()[0]
                product_dict['num_reviews'] = num_reviews
            except:
                pass
        ## write everything into the csv
        writer.writerow(product_dict)
        time.sleep(2)
        print("end")
except Exception as e:
    ## if there is an error with the page, print the url and error message so
    ## the page can be manually inspected
    print(url)
    print(e)
    ## write the index so the script knows where where it left off
    open('index.txt', 'w').write('%d' % (i + index))
    csv_file.close()
    driver.quit()
    time.sleep(5)
    quit()

driver.quit()
quit()

Scraper/get_products_list.py
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.by import By

import time
import os
import csv
import math

### This is a script that goes to each brand page and gets the urls for each
### product. It should be run second.
### The brand pages are set up such that in order to see the next page,
### you have to click the "next page" at the bottom of the screen
### Each product shows up in it's own "lazy loading" div, meaning that in
### order for it to be "seen" the driver must scroll down to that element

### One challenge was dealing with the page "freezing", especially when the
### number of products on the page was greater than about 40.  The
### window needs to be reactivated by clicking on an element. However clicking
### on an element usually results in a  pop-up. Thus, the order of events in
### the "scrape_brandpage_v1" function is to see if the elements are loading,
### and if they are not, to try to click on the last visible element and then
### pass PAGE DOWN keys to it. If there is an exception, then there is a
### popup, so the "continue shopping" button is clicked and then the PAGE
### DOWN keys are again passed to the element

### The pauses are really long because ironically "lazy loading" takes a long
### time and also this prevents the browser from crashing.

### Sometimes a pop-up would show up again after the second click, and then
### after clicking the "continue shopping" button, the page would
### redirect to the url of the product element that was clicked. This usually
### happens when the brand has two or more pages of products. Thus there is a
### "scrape_brandpage_v2" function which after removing the first popup, first
###  passes PAGE DOWN before clicking then removes the pop-up again, then
### clicks the element again.

### There's also scraping_brandpage_v3, which did not need any clicking or
### PAGE DOWNs after clicking the "continue shopping" button

### The change_page() function is for changing pages. This happens when
### clicking the "Next button" at the bottom of the page. If the driver tries
### to find elements straight away, there is a "stale element" error,
### however, refreshing the page results in the page returning to the first
### page of the brand. Thus the driver.execute_script("return document.\
### documentElement.outerHTML") command is run to get the elements on the page

### The maximum number of products per page is 60. Most brands have fewer than
### 60 products, and thus just use scrape_brandpage_v1. The ones with between
### 60 and 120 products use num_pages_2() function, which scrapes the second
### page with either v1 or v2. Only a few brands had more than 2 pages, with
### Clinique and Sephora being the only brands with more than 3 pages. These
### use a combination of v1, v2, and v3.
##############################################################################

## setting up the webdriver with a user-agent and with options to
## disable notifications - though it seems not to work well
opts = Options()
opts.add_argument("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36")
opts.add_argument("--disable-notifications")
driver = webdriver.Chrome(options=opts)

## open the brand page urls from the "get_brand_list.py" script
brands = open('sephora_brand_pages.txt', 'r').readlines()

## These are commonly used xpaths. The product_path is the path to the product
## urls the continue_shopping_xpath is the xpath to the "continue shopping"
## button on the pop-up modal dialogs that come up

product_xpath = '//div[@class="css-12egk0t"]/a[@class="css-ix8km1"]'
continue_shopping_xpath = "//button[@aria-label='Continue shopping']"


## the driver.execute_script() method executes javascript code
## window_scroll is the command for scrolling, with document.body.scrollHeight
## being the length of the window
## usually it is used as
## driver.execute_script("window.scrollTo(0, document.body.scrollHeight*x)")
## where x is some fraction between 0 and 1 to determine how far down the page
## to scroll. Thus window_scroll(x) will scroll to that point on the page

def window_scroll(x):
    return f"window.scrollTo(0, document.body.scrollHeight*{x})"


############################################################################
## This is a function for scraping brand pages for product urls
def scrape_brandpage_v1(url, num_products):
    driver.implicitly_wait(10)
    if url:
        driver.implicitly_wait(10)
        driver.get(url)
    if (num_products > 12) and (num_products <= 25):
        driver.execute_script(window_scroll(0.60))
        time.sleep(10)
    if (num_products > 25) and (num_products <= 36):
        driver.execute_script(window_scroll(0.6))
        time.sleep(10)
        driver.execute_script(window_scroll(0.4))
        time.sleep(10)
        driver.execute_script(window_scroll(0.8))
        time.sleep(10)
    if (num_products > 36):
        nproducts_initial = len(driver.find_elements_by_xpath(product_xpath))
        if nproducts_initial <= 36:
            driver.execute_script("arguments[0].scrollIntoView()",
                                  driver.find_elements_by_xpath(product_xpath)[-1])
            time.sleep(10)
            nproducts_final = len(driver.find_elements_by_xpath(product_xpath))
            if nproducts_initial == nproducts_final:
                try:
                    driver.find_elements_by_xpath(product_xpath)[-1]. \
                        send_keys(Keys.PAGE_DOWN)
                    ex_out = driver. \
                        find_element_by_xpath(continue_shopping_xpath)
                    wait = WebDriverWait(driver, 10)
                    element = wait.until(EC.element_to_be_clickable((By.XPATH, continue_shopping_xpath)))
                    ex_out.click()
                except:
                    time.sleep(10)
                    # driver.find_elements_by_xpath(product_xpath)[-1]. \
                    #     send_keys(Keys.PAGE_DOWN)
        time.sleep(5)
        driver.execute_script("arguments[0].scrollIntoView()",
                              driver.find_elements_by_xpath(product_xpath)[-1])
        time.sleep(10)
        driver.execute_script("arguments[0].scrollIntoView()", \
                              driver.find_elements_by_xpath(product_xpath)[-1])
        time.sleep(15)
        driver.execute_script(window_scroll(0.6))
        time.sleep(10)
        driver.execute_script("arguments[0].scrollIntoView()", \
                              driver.find_elements_by_xpath(product_xpath)[-1])
        time.sleep(10)
    products = driver.find_elements_by_xpath(product_xpath)
    product_urls = [product.get_attribute('href') for product in products]
    return product_urls


def scrape_brandpage_v2():
    driver.implicitly_wait(10)
    nproducts_initial = len(driver.find_elements_by_xpath(product_xpath))
    driver.execute_script("arguments[0].scrollIntoView()", driver.find_elements_by_xpath(product_xpath)[-1])
    time.sleep(10)
    nproducts_final = len(driver.find_elements_by_xpath(product_xpath))
    if nproducts_initial == nproducts_final:
        time.sleep(5)
        try:
            ex_out = driver.find_element_by_xpath(continue_shopping_xpath)
            wait = WebDriverWait(driver, 10)
            element = wait.until(EC.element_to_be_clickable((By.XPATH, continue_shopping_xpath)))
            ex_out.click()
        except:
            time.sleep(3)
            driver.execute_script(window_scroll(0.8))
            time.sleep(15)
        nproducts = len(driver.find_elements_by_xpath(product_xpath))
        if nproducts == nproducts_final:
            driver.find_elements_by_xpath(product_xpath)[5]. \
                send_keys(Keys.PAGE_DOWN)
            time.sleep(15)
            nproducts = len(driver.find_elements_by_xpath(product_xpath))
            time.sleep(5)
            try:
                ex_out = driver.find_element_by_xpath(continue_shopping_xpath)
                wait = WebDriverWait(driver, 10)
                element = wait.until(EC.element_to_be_clickable((By.XPATH, continue_shopping_xpath)))
                ex_out.click()
            except:
                time.sleep(0.5)
    driver.execute_script(window_scroll(0.2))
    time.sleep(2)
    driver.execute_script(window_scroll(0.5))
    time.sleep(15)
    driver.execute_script(window_scroll(0.8))
    time.sleep(5)
    driver.execute_script(window_scroll(0.615))
    time.sleep(15)
    products = driver.find_elements_by_xpath(product_xpath)
    product_urls = [product.get_attribute('href') for product in products]
    return product_urls


def scrape_brandpage_v3():
    nproducts_initial = len(driver.find_elements_by_xpath(product_xpath))
    driver.execute_script("arguments[0].scrollIntoView()", \
                          driver.find_elements_by_xpath(product_xpath)[-1])
    time.sleep(10)
    nproducts_final = len(driver.find_elements_by_xpath(product_xpath))
    if nproducts_initial == nproducts_final:
        try:
            ex_out = driver.find_element_by_xpath(continue_shopping_xpath)
            wait = WebDriverWait(driver, 10)
            element = wait.until(EC.element_to_be_clickable((By.XPATH, continue_shopping_xpath)))
            ex_out.click()
        except:
            time.sleep(3)
            driver.execute_script(window_scroll(0.8))
            time.sleep(15)
        nproducts = len(driver.find_elements_by_xpath(product_xpath))
    time.sleep(10)
    driver.execute_script(window_scroll(0.2))
    time.sleep(2)
    driver.execute_script(window_scroll(0.5))
    time.sleep(15)
    driver.execute_script(window_scroll(0.8))
    time.sleep(5)
    driver.execute_script(window_scroll(0.615))
    time.sleep(15)
    products = driver.find_elements_by_xpath(product_xpath)
    product_urls = [product.get_attribute('href') for product in products]
    return product_urls


def change_page():
    try:
        next_page_button = driver.find_element_by_xpath('//nav[@aria-label=' + \
                                                        '"Pagination"]//button[@aria-label="Next"]')
        next_page_button.click()
    except:
        ex_out = driver.find_element_by_xpath(continue_shopping_xpath)
        time.sleep(2)
        ex_out.click()
        time.sleep(2)
        next_page_button.click()
    time.sleep(2)
    driver.execute_script(window_scroll(0.60))
    time.sleep(4)
    driver.execute_script("return document.documentElement.outerHTML")
    time.sleep(15)


def num_pages_2(brand, num_tot_products, method):
    global brand_urls
    brand_urls= scrape_brandpage_v1(brand, 60)
    change_page()
    if method == 1:
        brand_urls.extend(scrape_brandpage_v1([], num_tot_products - 60))
    if method == 2:
        brand_urls.extend(scrape_brandpage_v2())
    return brand_urls


### these lines check to see if scrape has already started. Because of browser
### crashes or unexpected issues or accidentally closing the browser window,
### the script will save what it has in sephora_product_urls.txt and then save
### the index where it left off in index.txt, so that when rerun, the script
### will simply start at that index and append to the existing urls
if os.path.isfile('index.txt') and os.path.isfile('sephora_product_urls.txt'):
    index = int(open('index.txt').read())
    f = open('sephora_product_urls.txt', 'a')
else:
    index = 0
    f = open('sephora_product_urls.txt', 'w')

### This is for running the scraper, different procedures are used depending
### on how many product pages that brand has
try:
    for i, brand in enumerate(brands[index:]):
        driver.implicitly_wait(10)
        driver.get(brand)
        try:
            num_tot_products = int(driver.find_element_by_xpath('//span' + \
                                                                '[@data-at="number_of_products"]').text.split()[0])
            print(num_tot_products)
        except:
            num_tot_products = 0
            continue
        if num_tot_products <= 60:
            brand_urls = scrape_brandpage_v1(brand, num_tot_products)
        elif num_tot_products <= 120:
            try:
                brand_urls = num_pages_2(brand, num_tot_products, 1)
            except:
                num_pages_2(brand, num_tot_products, 2)
        elif num_tot_products <= 180:
            brand_urls = num_pages_2(brand, num_tot_products, 2)
            change_page()
            brand_urls.extend(scrape_brandpage_v2())
        else:
            num_pages = math.ceil(num_tot_products / 60) - 3
            num_last_page = num_tot_products % 60
            if num_last_page == 0:
                num_last_page = 60
            brand_urls = num_pages_2(brand, 120, 2)
            for _ in range(num_pages):
                change_page()
                brand_urls.extend(scrape_brandpage_v3())
            change_page()
            brand_urls.extend(scrape_brandpage_v2())

        for url in brand_urls:
            f.write(url + '\n')
except Exception as e:
    print(e)
    f.close()
    new_index = i + index
    open('index.txt', 'w').write('%d' % (i + index))

quit()
########################################################################

Scraper/get_brands_list.py
from selenium import webdriver
from selenium.webdriver.chrome.options import Options

### This is a script for getting the urls of the brand pages at sephora. 
### This should be run first

# Setting up the webdriver with a user agent and attempting to diable pop-ups
opts = Options()
opts.add_argument("user-agent=['Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36']")
opts.add_argument("--disable-notifications")
driver = webdriver.Chrome(options=opts)

# start at the page listing brands
start_url = "https://www.sephora.com/brands-list"
driver.get(start_url)

## getting the urls by their xpath
# brands = driver.find_elements_by_xpath('//div//ul/li/a')
brands = driver.find_elements_by_xpath('//a[@data-at="brand_link"]')
brands = [a.get_attribute('href')+'/all' for a in brands]

## saving the urls to a text file
with open('sephora_brand_pages.txt', 'w') as f:
	for url in brands:
		f.write(url+'\n')
f.close()
driver.close()
